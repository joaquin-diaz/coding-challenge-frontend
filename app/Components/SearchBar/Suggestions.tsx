import * as React from 'react';
import { consolidateSuggestions, Film } from '../../types/Film';
import Card from '../../Components/Card';
import styled from '@emotion/styled';
import styles from '../../constants/styles';

const Container = styled(Card)`
  width: 100%;
  padding: 10px 0;
`;

const Suggestion = styled.div`
  padding: 10px 15px;
  cursor: pointer;

  &:hover {
    background-color: ${styles.colors.gray};
  }
`;

const Match = styled.div`
  font-size: 12px;
  font-style: italic;
`;

const { useState, useEffect } = React;

interface SuggestionsProps {
  suggestedFilms: Array<Film>;
  query: string;
  onSuggestionClick: (title: string) => void;
}

const Suggestions: React.FC<SuggestionsProps> = props => {
  const [suggestions, setSuggestions] = useState<Array<string>>([]);
  useEffect(() => {
    const s = consolidateSuggestions(props.suggestedFilms, props.query);
    setSuggestions(s);
  }, [props.suggestedFilms.length]);

  const renderSuggestion = (suggestion: string, key: number) => {
    const [title, match] = suggestion.split('//');
    const handleClick = () => {
      props.onSuggestionClick(title);
    };

    return (
      <Suggestion key={key} onClick={handleClick}>
        <div>{title}</div>
        <Match>{match}</Match>
      </Suggestion>
    );
  };

  if (suggestions.length === 0) return null;

  return <Container>{suggestions.map(renderSuggestion)}</Container>;
};

export default Suggestions;
