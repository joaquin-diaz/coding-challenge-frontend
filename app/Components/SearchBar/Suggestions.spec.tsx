import * as React from 'react';

import Suggestions from './Suggestions';
import { mount } from 'enzyme';

const suggestions = [
  {
    title: 'Something',
    subtitle: 'Subsomething',
  },
  {
    title: 'Something',
    subtitle: 'Subsomething',
  },
];

const noop = () => {};

describe('<Suggestions />', () => {
  it('should render', () => {
    const wrapper = mount(<Suggestions suggestions={suggestions} onSuggestionClick={noop} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should be null', () => {
    const wrapper = mount(<Suggestions suggestions={[]} onSuggestionClick={noop} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call on click suggestion', () => {
    const mockSuggestionClick = jest.fn();

    const wrapper = mount(
      <Suggestions suggestions={suggestions} onSuggestionClick={mockSuggestionClick} />,
    );
    const suggestionRow = wrapper.find('div').at(2);

    expect(suggestionRow.exists).toBeTruthy();
    suggestionRow.simulate('click');

    expect(mockSuggestionClick).toBeCalled();
  });
});
