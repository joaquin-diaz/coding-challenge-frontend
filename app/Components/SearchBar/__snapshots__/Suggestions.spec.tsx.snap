// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Suggestions /> should be null 1`] = `
<Suggestions
  onSuggestionClick={[Function]}
  suggestions={Array []}
/>
`;

exports[`<Suggestions /> should render 1`] = `
.emotion-4 {
  background: white;
  box-shadow: 0 1px 6px 0 rgba(32,33,36,0.28);
  box-sizing: border-box;
  border-radius: 5px;
  width: 100%;
  padding: 10px 0;
}

.emotion-1 {
  padding: 10px 15px;
  cursor: pointer;
}

.emotion-1:hover {
  background-color: #eceff1;
}

.emotion-0 {
  font-size: 12px;
  font-style: italic;
}

<Suggestions
  onSuggestionClick={[Function]}
  suggestions={
    Array [
      Object {
        "subtitle": "Subsomething",
        "title": "Something",
      },
      Object {
        "subtitle": "Subsomething",
        "title": "Something",
      },
    ]
  }
>
  <Styled(div)>
    <div
      className="emotion-4"
    >
      <Styled(div)
        key="0"
        onClick={[Function]}
      >
        <div
          className="emotion-1"
          onClick={[Function]}
        >
          <div>
            Something
          </div>
          <Styled(div)>
            <div
              className="emotion-0"
            >
              Subsomething
            </div>
          </Styled(div)>
        </div>
      </Styled(div)>
      <Styled(div)
        key="1"
        onClick={[Function]}
      >
        <div
          className="emotion-1"
          onClick={[Function]}
        >
          <div>
            Something
          </div>
          <Styled(div)>
            <div
              className="emotion-0"
            >
              Subsomething
            </div>
          </Styled(div)>
        </div>
      </Styled(div)>
    </div>
  </Styled(div)>
</Suggestions>
`;
